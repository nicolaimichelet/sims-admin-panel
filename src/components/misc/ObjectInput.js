import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Toggle from 'material-ui/Toggle';
import { mapAndConnect, IManagedService, ManagedService} from "services";
import {Redirect} from "react-router";

import DropDownMenu from 'material-ui/DropDownMenu';
import MenuItem from 'material-ui/MenuItem';


import { get, set, has, cloneDeep } from 'lodash';
import { paths } from 'common/utils';

import { assign, keys, union } from 'lodash';

export class ObjectInput extends Component{
  constructor(props){
    super(props);
    this.state = {
      value: {}
    }
  }

  componentDidMount(){
    this._updateFields(this.props);
  }

  componentWillReceiveProps(nextProps){
    this._updateFields(nextProps);
  }

  shouldComponentUpdate(nextProps, nextState){
    let nextPropVal = nextProps.value || {};
    let oldPropVal = this.props.value || {};
    let newKeys = union(paths(nextPropVal), paths(nextState.value || {}));
    let oldKeys = union(paths(oldPropVal), paths(this.state.value || {}));

    

    if(newKeys.length != oldKeys.length){
      return true;
    }

    let allKeys = union(newKeys, oldKeys);

    for(let key of allKeys){
      if(get(nextState.value, key) !== get(this.state.value, key)){
        return true;
      }
    }

    for(let key of allKeys){
      if(get(nextPropVal, key) !== get(oldPropVal, key)){
        return true;
      }
    }



    return false;
  }

  isControlled(){
    return this.props.value != null;
  }

  _updateFields(props){
    let nmap = {};
    if(props.value != null){
      for(let atr of paths(props.value) ){
        set(nmap, atr, get(props.value, atr));
      }
      this.setState({
        value: nmap
      });
    }
  }

  onFieldChange(name, nval){
    let object = cloneDeep(this.state.value);
    set(object, name, nval);
    if(this.props.map) {
      object = this.props.map(object);
    }

    this.setState({
      value: object
    }, () => {
      if(this.props.onChange){
        this.props.onChange(object);
      }
    });
    
    


  }


  render(){

    const inputs = React.Children.map(this.props.children, (field, idx) => {
      if(field.props.name){
        let onChange = (...a) => {
          let value = a[1];
          if(field.props.mapValue){
            value = field.props.mapValue(...a);
          }else if(this.props.mapValues){
            value = this.props.mapValues(...a);
          }

          this.onFieldChange(field.props.name, value);
          if(field.props.onChange){
            field.props.onChange(...a);
          }
          
          if(field.props.onCheck){
            field.props.onCheck(...a);
          }

        }
        return React.cloneElement(field, {
          onChange: onChange,
          onCheck: onChange,
          value: this.isControlled() ? get(this.state.value, field.props.name) : undefined,
          checked: this.isControlled() ? get(this.state.value, field.props.name) : undefined,
          key: field.key || field.props.key || idx
        });
      }

      return field;
    });
    return (
      <div style={{display: "inline-block"}}>
        {inputs}
      </div>
    );
  }
}


ObjectInput.defaultProps = {
  mapValues: (event, v) => {
    if(v != null){return v;}
    return event.target.value || event.target.checked;
  }
}

ObjectInput.propTypes = {
  value: PropTypes.object,
  mapValues: PropTypes.func
}

