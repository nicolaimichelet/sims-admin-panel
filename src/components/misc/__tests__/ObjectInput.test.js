import React from 'react';
import ReactDOM from 'react-dom';
import { ObjectInput } from 'components/misc/ObjectInput.js';

import { mount, shallow, render } from 'enzyme';
import { wrap } from 'module';

describe('ObjectInput', () => {
  
  it('renders with no props', () => {
    const wrapper = shallow(<ObjectInput />);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders with inputs', () => {
    const wrapper = shallow(
      <ObjectInput>
        <input name="name" type="text" />
        <input name="href" type="text" />
      </ObjectInput>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('renders with value and inputs', () => {

    const formData = {
      name: "server",
      href: "http://server:port"
    }
    const wrapper = shallow(
      <ObjectInput value={formData}>
        <input name="name" type="text" />
        <input name="href" type="text" />
      </ObjectInput>
    );

    expect(wrapper.find('[name="name"]').props().value).toBe(formData.name);
    expect(wrapper.find('[name="href"]').props().value).toBe(formData.href);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders with sub objects', () => {
    const formData = {
      name: "server",
      href: "http://server:port",
      spec: {
        type: "speed",
        value: "100Mbps"
      }
    }
    const wrapper = shallow(
      <ObjectInput value={formData}>
        <input name="name" type="text" />
        <input name="href" type="text" />
        <input name="spec.type" type="text" />
        <input name="spec.value" type="text" />
      </ObjectInput>
    );

    expect(wrapper.find('[name="name"]').props().value).toBe(formData.name);
    expect(wrapper.find('[name="href"]').props().value).toBe(formData.href);
    
    expect(wrapper.find('[name="spec.type"]').props().value).toBe(formData.spec.type);
    expect(wrapper.find('[name="spec.value"]').props().value).toBe(formData.spec.value);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders with nested ObjectInputs', () => {
    const formData = {
      name: "server",
      href: "http://server:port",
      spec: {
        type: "speed",
        value: "100Mbps"
      }
    }

    const wrapper = mount(
      <ObjectInput value={formData}>
        <input name="name" type="text" />
        <input name="href" type="text" />
        <ObjectInput name="spec">
          <input mapValue={(a)=>a} name="type" type="text" />
          <input onChange={jest.fn()} name="value" type="text" />
        </ObjectInput>
      </ObjectInput>
    );

    expect(wrapper.find('[name="name"]').props().value).toBe(formData.name);
    expect(wrapper.find('[name="href"]').props().value).toBe(formData.href);
    expect(wrapper.find('[name="spec"]').find('[name="type"]').props().value).toBe(formData.spec.type);
    expect(wrapper.find('[name="spec"]').find('[name="value"]').props().value).toBe(formData.spec.value);

  });

  it('mounts, changes props, simulates change and unmounts', () => {
    const formData = {
      name: "server",
      href: "http://server:port",
      spec: {
        type: "enabled",
        value: true
      }
    };

    const wrapper = mount(
      <ObjectInput value={formData}>
        <input name="name" type="text" />
        <input name="href" type="text" />
        <ObjectInput name="spec">
          <input name="type" type="text" />
          <input onCheck={jest.fn()} name="value" type="checkbox" />
        </ObjectInput>
      </ObjectInput>
    );

    formData.name = "notserver";
    wrapper.setProps({
      value: formData,
      onChange: jest.fn()
    });

    wrapper.update();

    expect(wrapper.find('[name="name"]').props().value).toBe(formData.name);
    expect(wrapper.find('[name="href"]').props().value).toBe(formData.href);
    
    const inner = wrapper.find('[name="spec"]');
    expect(inner.find('[name="type"]').props().value).toBe(formData.spec.type);
    expect(inner.find('[name="value"]').props().value).toBe(formData.spec.value);

    //wrapper.find('[name="value"]').simulate("change", "1234");
    //wrapper.find('[name="href"]').simulate("change", "5678");

    //expect(wrapper.props.onChange).toHaveBeenCalled();

    wrapper.unmount();
  });


});
