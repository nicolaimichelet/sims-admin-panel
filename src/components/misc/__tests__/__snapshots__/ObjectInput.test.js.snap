// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ObjectInput renders with inputs 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ObjectInput
    mapValues={[Function]}
>
    <input
        name="name"
        type="text"
    />
    <input
        name="href"
        type="text"
    />
</ObjectInput>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <input
          checked={undefined}
          name="name"
          onChange={[Function]}
          onCheck={[Function]}
          type="text"
          value={undefined}
/>,
        <input
          checked={undefined}
          name="href"
          onChange={[Function]}
          onCheck={[Function]}
          type="text"
          value={undefined}
/>,
      ],
      "style": Object {
        "display": "inline-block",
      },
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "0/.0",
        "nodeType": "host",
        "props": Object {
          "checked": undefined,
          "name": "name",
          "onChange": [Function],
          "onCheck": [Function],
          "type": "text",
          "value": undefined,
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": "1/.1",
        "nodeType": "host",
        "props": Object {
          "checked": undefined,
          "name": "href",
          "onChange": [Function],
          "onCheck": [Function],
          "type": "text",
          "value": undefined,
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <input
            checked={undefined}
            name="name"
            onChange={[Function]}
            onCheck={[Function]}
            type="text"
            value={undefined}
/>,
          <input
            checked={undefined}
            name="href"
            onChange={[Function]}
            onCheck={[Function]}
            type="text"
            value={undefined}
/>,
        ],
        "style": Object {
          "display": "inline-block",
        },
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "0/.0",
          "nodeType": "host",
          "props": Object {
            "checked": undefined,
            "name": "name",
            "onChange": [Function],
            "onCheck": [Function],
            "type": "text",
            "value": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": "1/.1",
          "nodeType": "host",
          "props": Object {
            "checked": undefined,
            "name": "href",
            "onChange": [Function],
            "onCheck": [Function],
            "type": "text",
            "value": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`ObjectInput renders with no props 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ObjectInput
    mapValues={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": undefined,
      "style": Object {
        "display": "inline-block",
      },
    },
    "ref": null,
    "rendered": null,
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": undefined,
        "style": Object {
          "display": "inline-block",
        },
      },
      "ref": null,
      "rendered": null,
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`ObjectInput renders with sub objects 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ObjectInput
    mapValues={[Function]}
    value={
        Object {
            "href": "http://server:port",
            "name": "server",
            "spec": Object {
              "type": "speed",
              "value": "100Mbps",
            },
          }
    }
>
    <input
        name="name"
        type="text"
    />
    <input
        name="href"
        type="text"
    />
    <input
        name="spec.type"
        type="text"
    />
    <input
        name="spec.value"
        type="text"
    />
</ObjectInput>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <input
          checked="server"
          name="name"
          onChange={[Function]}
          onCheck={[Function]}
          type="text"
          value="server"
/>,
        <input
          checked="http://server:port"
          name="href"
          onChange={[Function]}
          onCheck={[Function]}
          type="text"
          value="http://server:port"
/>,
        <input
          checked="speed"
          name="spec.type"
          onChange={[Function]}
          onCheck={[Function]}
          type="text"
          value="speed"
/>,
        <input
          checked="100Mbps"
          name="spec.value"
          onChange={[Function]}
          onCheck={[Function]}
          type="text"
          value="100Mbps"
/>,
      ],
      "style": Object {
        "display": "inline-block",
      },
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "0/.0",
        "nodeType": "host",
        "props": Object {
          "checked": "server",
          "name": "name",
          "onChange": [Function],
          "onCheck": [Function],
          "type": "text",
          "value": "server",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": "1/.1",
        "nodeType": "host",
        "props": Object {
          "checked": "http://server:port",
          "name": "href",
          "onChange": [Function],
          "onCheck": [Function],
          "type": "text",
          "value": "http://server:port",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": "2/.2",
        "nodeType": "host",
        "props": Object {
          "checked": "speed",
          "name": "spec.type",
          "onChange": [Function],
          "onCheck": [Function],
          "type": "text",
          "value": "speed",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": "3/.3",
        "nodeType": "host",
        "props": Object {
          "checked": "100Mbps",
          "name": "spec.value",
          "onChange": [Function],
          "onCheck": [Function],
          "type": "text",
          "value": "100Mbps",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <input
            checked="server"
            name="name"
            onChange={[Function]}
            onCheck={[Function]}
            type="text"
            value="server"
/>,
          <input
            checked="http://server:port"
            name="href"
            onChange={[Function]}
            onCheck={[Function]}
            type="text"
            value="http://server:port"
/>,
          <input
            checked="speed"
            name="spec.type"
            onChange={[Function]}
            onCheck={[Function]}
            type="text"
            value="speed"
/>,
          <input
            checked="100Mbps"
            name="spec.value"
            onChange={[Function]}
            onCheck={[Function]}
            type="text"
            value="100Mbps"
/>,
        ],
        "style": Object {
          "display": "inline-block",
        },
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "0/.0",
          "nodeType": "host",
          "props": Object {
            "checked": "server",
            "name": "name",
            "onChange": [Function],
            "onCheck": [Function],
            "type": "text",
            "value": "server",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": "1/.1",
          "nodeType": "host",
          "props": Object {
            "checked": "http://server:port",
            "name": "href",
            "onChange": [Function],
            "onCheck": [Function],
            "type": "text",
            "value": "http://server:port",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": "2/.2",
          "nodeType": "host",
          "props": Object {
            "checked": "speed",
            "name": "spec.type",
            "onChange": [Function],
            "onCheck": [Function],
            "type": "text",
            "value": "speed",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": "3/.3",
          "nodeType": "host",
          "props": Object {
            "checked": "100Mbps",
            "name": "spec.value",
            "onChange": [Function],
            "onCheck": [Function],
            "type": "text",
            "value": "100Mbps",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`ObjectInput renders with value and inputs 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ObjectInput
    mapValues={[Function]}
    value={
        Object {
            "href": "http://server:port",
            "name": "server",
          }
    }
>
    <input
        name="name"
        type="text"
    />
    <input
        name="href"
        type="text"
    />
</ObjectInput>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <input
          checked="server"
          name="name"
          onChange={[Function]}
          onCheck={[Function]}
          type="text"
          value="server"
/>,
        <input
          checked="http://server:port"
          name="href"
          onChange={[Function]}
          onCheck={[Function]}
          type="text"
          value="http://server:port"
/>,
      ],
      "style": Object {
        "display": "inline-block",
      },
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "0/.0",
        "nodeType": "host",
        "props": Object {
          "checked": "server",
          "name": "name",
          "onChange": [Function],
          "onCheck": [Function],
          "type": "text",
          "value": "server",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": "1/.1",
        "nodeType": "host",
        "props": Object {
          "checked": "http://server:port",
          "name": "href",
          "onChange": [Function],
          "onCheck": [Function],
          "type": "text",
          "value": "http://server:port",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <input
            checked="server"
            name="name"
            onChange={[Function]}
            onCheck={[Function]}
            type="text"
            value="server"
/>,
          <input
            checked="http://server:port"
            name="href"
            onChange={[Function]}
            onCheck={[Function]}
            type="text"
            value="http://server:port"
/>,
        ],
        "style": Object {
          "display": "inline-block",
        },
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "0/.0",
          "nodeType": "host",
          "props": Object {
            "checked": "server",
            "name": "name",
            "onChange": [Function],
            "onCheck": [Function],
            "type": "text",
            "value": "server",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": "1/.1",
          "nodeType": "host",
          "props": Object {
            "checked": "http://server:port",
            "name": "href",
            "onChange": [Function],
            "onCheck": [Function],
            "type": "text",
            "value": "http://server:port",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
